specification/dartLangSpec.tex

```diff
@@ -11716,18 +11720,38 @@ \subsection{Function Expressions}
 }
 
 \LMHash{}%
-We say that $S$ is the
-\IndexCustom{future type derived from a type}{type!future type derived from}
-$T$ in the following cases, using the first applicable case:
+We say that a type $T$
+\IndexCustom{derives a future type}{type!derives a future type}
+$F$ in the following cases, using the first applicable case:
 
+%% TODO(eernst): Note that `X extends X?` can create an infinite loop.
+%% We will probably make that kind of bound an error.
 \begin{itemize}
-\item $T$ implements $S$
-  (\ref{interfaceSuperinterfaces}),
-  and there is a $U$ such that $S$ is \code{Future<$U$>}.
-\item $T$ is $S$ bounded
-  (\ref{bindingActualsToFormals}),
-  and there is a $U$ such that
-  $S$ is \code{FutureOr<$U$>}, \code{Future<$U$>?}, or \code{FutureOr<$U$>?}.
+\item
+  %% TODO(eernst): Come mixin classes and extension types: add them.
+  If $T$ is a type which is introduced by
+  a class, mixin, or enum declaration,
+  and if $T$ or a direct or indirect superinterface
+  (\ref{interfaceSuperinterfaces})
+  of $T$ is \code{Future<$U$>} for some $U$,
+  then $T$ derives the future type \code{Future<$U$>}.
+\item
+  If $T$ is the type \code{FutureOr<$U$>} for some $U$,
+  then $T$ derives the future type \code{FutureOr<$U$>}.
+\item
+  If $T$ is \code{$S$?} for some $S$, and
+  $S$ derives the future type $F$,
+  then $T$ derives the future type \code{$F$?}.
+\item
+  If $T$ is a type variable with bound $B$, and
+  $B$ derives the future type $F$,
+  then $T$ derives the future type $F$.
+\item
+  \commentary{%
+    There is no rule for the case where $T$ is of the form \code{$X$\,\&\,$S$}
+    because this will never occur
+    (this concept is only used in \flattenName, which is defined below).%
```

Comment:
If I understand it correctly, the concept is also used in the definition of types incompatible with `await`, where the right-hand side of an intersection can be tested for deriving a future. If it does, the entire intersection type is deemed not incompatible with `await`. Consider the following example.

```dart
foo<X extends Future<int>?, Y extends Future<int>?>(X x) async {
  if (x is Y) {
    await (x ?? x!);
  }
}
```

The static type of the awaited expression `x ?? x!` is `X & (Y & Future<int>)`. To test if it's incompatible with `await`, we look at the right-hand side of the intersection, which is `Y & Future<int>` and see if it derives a future type. Intuitively it seems that it does. So it looks like the rule for intersection type should be defined.

Reply:
UPDATE: The example works with `await x!` too. The static type of `x!` is the same nested intersection type `X & (Y & Future<int>)`. I used `(x ?? x!)` in the original example because it provides a visible representation of the static type of the awaited expression as shown in the following snippet.

```
library;
import self as self;
import "dart:async" as asy;
import "dart:core" as core;

static method foo<X extends asy::Future<core::int>?, Y extends asy::Future<core::int>?>(self::foo::X% x) → dynamic async /* emittedValueType= dynamic */ {
  if(x is self::foo::Y%) {
    await let final self::foo::X% & self::foo::Y% /* '%' & '%' = '%' */ #t1 = x{self::foo::X% & self::foo::Y% /* '%' & '%' = '%' */} in #t1 == null ?{self::foo::X% & self::foo::Y% & asy::Future<core::int> /* '%' & '!' = '!' */ /* '%' & '!' = '!' */} x{self::foo::X% & self::foo::Y% /* '%' & '%' = '%' */}! : #t1{self::foo::X% & self::foo::Y% & asy::Future<core::int> /* '%' & '!' = '!' */ /* '%' & '!' = '!' */};
  }
}
```